draft-ietf-tls-esni.md

```diff
@@ -411,17 +414,42 @@ any PSK binders, and then MAY substitute extensions which it knows will be
 duplicated in ClientHelloOuter. To do so, the client computes a hash of the
 entire ClientHelloInner message with the same hash as for the KDF used to
 encrypt ClientHelloInner. Then, the client removes and replaces extensions from
-ClientHelloInner with a single "outer_extensions" extension. The list of
-`outer_extensions` include those which were removed from ClientHelloInner, in
-the order in which they were removed. The hash contains the full
+ClientHelloInner with a single "ech_outer_extensions" extension. The list of
+`ech_outer_extensions` include those which were removed from ClientHelloInner,
+in the order in which they were removed. The hash contains the full
 ClientHelloInner hash computed above.
 
 This process is reversed by client-facing server. Specifically,
-the server replaces the `outer_extensions` with extensions contained in
+the server replaces the `ech_outer_extensions` with extensions contained in
 ClientHelloOuter. The server then computes a hash of the reconstructed
 ClientHelloInner. If the hash does not equal OuterExtensions.hash, the server
 aborts the connection with an "illegal_parameter" alert.
 
+# The "ech_confirm" Extension {#ech-confirm}
+
+This extension allows the client to request from the server an explicit
+confirmation of ECH acceptance. When the client offers an "ech" extension, it
+MAY also offer an empty "ech_confirm" extension in its ClientHelloOuter. It MUST
```

Comment:
Why is this optional? Either it's part of the protocol or not. The spec should pick one. Variations in specs cause complexity throughout the system (every server would need to test both and make sure they work) and should only be there when they're necessary.

Note also that TLS does not allow unsolicited extensions, so any ServerHello extension either needs to be in the ClientHello (but mandated), or we that if you offer `ech`, an `ech_confirm` response is allowed as a special case. Alternatively...

Having two separate extensions at all seems unnecessary. Why not just use the `ech` codepoint? That would avoid this entire mess. There's plenty of precedent for extensions taking different forms in different contexts (`key_share` looks different between ClientHello, HelloRetryRequest, and ServerHello).

(Hrm. I'm not sure how the unsolicited extensions rule works if the inner ClientHello is chosen and lacks the `ech` extension. Something we may also need to sort out.)

Reply:
I had the same thought. The problem I see is that we currently use the "ech" codepoint for retry keys. We might define a codepoint for "ech_retry" so that the confirmation logic is carried by "ech". However, the client needs to solicit this extension.

I'm open to this if it makes more sense!

Reply:
The retry keys are carried in EncryptedExtensions while this is ServerHello, right? I don't think we've any precedent yet for an SH + EE extension, but extensions having different format between messages is common and seems fine to me.

Reply:
One nice thing about "ech" / "ech_retry" is that aligns ECH with the usual TLS extension logic, since the server can opt-out of both without breaking the protocol. Currently if the client offers "ech_confirm", the server MUST respond, or else be non-compliant.

Reply:
> The retry keys are carried in EncryptedExtensions while this is ServerHello, right?

Yes.

> I don't think we've any precedent yet for an SH + EE extension, but extensions having different format between messages is common and seems fine to me.

I was under the impression that this is disallowed in TLS 1.3, but looking at RFC 8446 I'm not so sure.

Reply:
> I was under the impression that this is disallowed in TLS 1.3, but looking at RFC 8446 I'm not so sure.

To clarify: I'm aware of extensions in which the server's response may be in one of many messages, but I'm not aware of any case in which the same extension appears in multiple messages in the same flight. I'd be fine with this, though it might be a pain for NSS (cc/ @martinthomson).

Reply:
I think that we can make it work if we need to.

Reply:
Right on, I'll update this PR tomorrow morning.

Reply:
Updated! I dropped "ech_confirm" and changed the semantics of "ech". Specifically, the extension appears in four messageas:
- ClientHelloOuter, same payload as usual;
- CleintHelloInner, with a random ECHNonce;
- ServerHello, with an ECHNonce that is the same as the ClientHelloInner in case of acceptance and a random value otherwise; and
- EncryptedClientHello, with retry ECHConfigs.